### PR TITLE
Filters Task (Security)

### PR DESCRIPTION
**Pull request by IvanLuchkin**
- implemented and applied AuthenticationFilter
- implemented LoginController and RegistrationController
- corresponding .jsp pages changed
- AuthenticationService implemented

### CODE REVIEW

File: src/main/java/taxi/model/Driver.java [hunk 1]
```diff
@@ -6,12 +6,30 @@
...
+    }
+
+    public String getPassword() {
+        return password;
+    }
+
+    public void setPassword(String password) {
+        this.password = password;
+    }
+
```
Comment (Max-Pochepets) [inline src/main/java/taxi/model/Driver.java:32]: if we would have to compare two drivers with different login/password, what will we get?

File: src/main/java/taxi/service/impl/CarServiceImpl.java [hunk 2]
```diff
@@ -22,7 +23,7 @@ public Car create(Car car) {
     @Override
     public Car get(Long id) {
         return carDao.get(id).orElseThrow(() ->
-                new RuntimeException("Car with id " + id + " not found"));
+                new NoSuchElementException("Car with id " + id + " not found"));
```
Comment (Max-Pochepets) [inline src/main/java/taxi/service/impl/CarServiceImpl.java:26]: just a suggestion: maybe it's better not to catch runtime exceptions but to process Optional with `isPresent`?

File: src/main/java/taxi/web/filters/AuthenticationFilter.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        HttpServletResponse resp = (HttpServletResponse) response;
+
+        String url = req.getServletPath();
+        if (url.equals("/login") || url.equals("/register")) {
+            filterChain.doFilter(req, resp);
+            return;
+        }
+
+        Long driverId = (Long) req.getSession().getAttribute(DRIVER_ID);
+        if (driverId == null) {
```
Comment (Max-Pochepets) [inline src/main/java/taxi/web/filters/AuthenticationFilter.java:37]: what will happen if logged in user deletes himself from the list? (your idea)

File: src/main/java/taxi/web/servlets/LoginController.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        String login = req.getParameter("login");
+        String password = req.getParameter("password");
+
+        try {
+            Driver driver = authenticationService.login(login, password);
+            req.getSession().setAttribute("driver_id", driver.getId());
+        } catch (AuthenticationException e) {
+            req.setAttribute("message", e.getMessage());
+            req.getRequestDispatcher("/WEB-INF/views/login.jsp").forward(req, resp);
+            return;
```
Comment (Max-Pochepets) [inline src/main/java/taxi/web/servlets/LoginController.java:35]:
```suggestion
```

**Review by Max-Pochepets — APPROVED** [review]

File: src/main/java/taxi/service/impl/CarServiceImpl.java (hunk 2 above)
Comment (IvanLuchkin) [reply src/main/java/taxi/service/impl/CarServiceImpl.java:26]: How?

File: src/main/java/taxi/web/filters/AuthenticationFilter.java (hunk 3 above)
Comment (IvanLuchkin) [reply src/main/java/taxi/web/filters/AuthenticationFilter.java:37]: Then `.get()` method will throw an exception

File: src/main/java/taxi/web/servlets/LoginController.java (hunk 4 above)
Comment (IvanLuchkin) [reply src/main/java/taxi/web/servlets/LoginController.java:35]: Why?

File: src/main/java/taxi/service/impl/CarServiceImpl.java (hunk 2 above)
Comment (Max-Pochepets) [reply src/main/java/taxi/service/impl/CarServiceImpl.java:26]: In service

File: src/main/java/taxi/web/filters/AuthenticationFilter.java (hunk 3 above)
Comment (Max-Pochepets) [reply src/main/java/taxi/web/filters/AuthenticationFilter.java:37]: You can process an Optional with a `isPresent` in service, returning null if no value present. I know that returning null is a bad practice, but it's better to process null value than show an exception page to the user

File: src/main/java/taxi/web/servlets/LoginController.java (hunk 4 above)
Comment (Max-Pochepets) [reply src/main/java/taxi/web/servlets/LoginController.java:35]: And why do you need it? Does this return change smth?

Comment (Max-Pochepets) [reply src/main/java/taxi/web/servlets/LoginController.java:35]: Nvm, found an explanation article, it's ok 👍

File: src/main/java/taxi/security/AuthenticationServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+import taxi.exception.AuthenticationException;
+import taxi.injections.Inject;
+import taxi.injections.Service;
+import taxi.model.Driver;
+import taxi.service.DriverService;
+
+@Service
+public class AuthenticationServiceImpl implements AuthenticationService {
+    @Inject
+    DriverService driverService;
```
Comment (RomanDubovskyi) [inline src/main/java/taxi/security/AuthenticationServiceImpl.java]: what about access modifier

File: src/main/java/taxi/security/AuthenticationServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+    @Override
+    public Driver login(String login, String password) throws AuthenticationException {
+        Driver driver = driverService.findByLogin(login).orElseThrow(() ->
+                new AuthenticationException("Wrong username or password"));
+        if (driver.getPassword().equals(password)) {
+            return driver;
+        }
+        throw new AuthenticationException("Wrong username or password");
```
Comment (RomanDubovskyi) [inline src/main/java/taxi/security/AuthenticationServiceImpl.java]:
```suggestion
        Optional<Driver> driver = driverService.findByLogin(login);
        if (driver.isPresent() && driver.get().getPassword().equals(password)) {
            return driver.get();
        }
        throw new AuthenticationException("Wrong username or password");
```

File: src/main/java/taxi/web/filters/AuthenticationFilter.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    private static final String DRIVER_ID = "driver_id";
+
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain filterChain)
+            throws IOException, ServletException {
+        HttpServletRequest req = (HttpServletRequest) request;
+        HttpServletResponse resp = (HttpServletResponse) response;
+
+        String url = req.getServletPath();
+        if (url.equals("/login") || url.equals("/register")) {
```
Comment (RomanDubovskyi) [inline src/main/java/taxi/web/filters/AuthenticationFilter.java]: it's better to create a Set with allowed URLs, so if this list gets bigger than 2 elements it won't create complicated if-else construction

File: src/main/java/taxi/dao/jdbc/CarDaoJdbc.java [hunk 8]
```diff
@@ -123,8 +125,12 @@ private Driver createDriverInstance(ResultSet resultSet) throws SQLException {
         Long driverId = resultSet.getObject("d_id", Long.class);
         String name = resultSet.getObject("d_name", String.class);
         String licenseNumber = resultSet.getObject("license_number", String.class);
+        String login = resultSet.getObject("login", String.class);
+        String password = resultSet.getObject("password", String.class);
```
Comment (RomanDubovskyi) [inline src/main/java/taxi/dao/jdbc/CarDaoJdbc.java:129]: let's not show users' password when we fetch data from db

**Review by RomanDubovskyi — CHANGES_REQUESTED** [review]

File: src/main/java/taxi/dao/jdbc/CarDaoJdbc.java (hunk 8 above)
Comment (IvanLuchkin) [reply src/main/java/taxi/dao/jdbc/CarDaoJdbc.java:129]: It is used only for construction of Driver instance, it is not used in toString() or in .jsp files

Comment (RomanDubovskyi) [reply src/main/java/taxi/dao/jdbc/CarDaoJdbc.java:129]: In the future, you will use DTOs but in generally there is no point of passing password from backend to frontEnd
(for now it's optional to change it)

File: src/main/java/taxi/web/filters/AuthenticationFilter.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public class AuthenticationFilter implements Filter {
+    private static final String DRIVER_ID = "driver_id";
+    private static final Set<String> ALLOWED_URLS = new HashSet<>();
```
Comment (RomanDubovskyi) [inline src/main/java/taxi/web/filters/AuthenticationFilter.java]:
```suggestion
    private static final Set<String> allowedUrls = new HashSet<>();
```
it's not "deeply immutable" so we don't use constant naming

**Review by RomanDubovskyi — APPROVED** [review]
fix allowedUrls naming, rest is good

**Pull request merged**